### ISSUE TITLE
send `closing_signed` twice

### ISSUE THREAD

**Issue by nayuta-ueno** [ElementsProject/lightning#1361]
commit: 957513666c494e35f71cb735eb82faed08f7b594

c-lightning send `closing_signed` twice(below `2` and `3a`).
Our node send `closing_signed` same as `2`'s `fee_satoshis`, and c-lightning disconnect because our `fee_satoshis` isn't "strictly between".
Is it necessary to wait for the second `closing_singed` ?

----
* funder: our node
* fundee: c-lightning

0. (`shutdown` exchanged)
1. our node: send `closing_signed` (fee_satoshis: feeA)
2. c-lightning: send `closing_signed` (fee_satoshis: feeB (<feeA) )
3. message cross
  3a. c-lightning: send `closing_signed`(fee_satoshis: feeC (>feeB) )
  3b. our node: send `closing_signed`(fee_satoshis: feeB)
4. c-lightning send `error` and disconnect( remote offer `feeB` not between `feeC` and `feeA` )
----

```
[Wed Apr 11 17:22:49.455 2018] 2018-04-11T08:22:48.869Z lightningd(123264): lightning_channeld-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: peer_in WIRE_SHUTDOWN
[Wed Apr 11 17:22:49.455 2018] 2018-04-11T08:22:48.870Z lightning_gossipd(123272): Received channel_update for channel 1292289:8:0(0) now DISABLED
[Wed Apr 11 17:22:49.455 2018] 2018-04-11T08:22:48.870Z lightningd(123264): lightning_channeld-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: UPDATE WIRE_CHANNEL_GOT_SHUTDOWN
[Wed Apr 11 17:22:49.455 2018] 2018-04-11T08:22:48.870Z lightningd(123264): 03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: State changed from CHANNELD_NORMAL to CHANNELD_SHUTTING_DOWN
[Wed Apr 11 17:22:49.471 2018] 2018-04-11T08:22:48.885Z lightning_gossipd(123272): Received channel_update for channel 1292289:8:0(0) now DISABLED
[Wed Apr 11 17:22:49.478 2018] 2018-04-11T08:22:48.885Z lightningd(123264): lightning_channeld-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Trying commit
[Wed Apr 11 17:22:49.478 2018] 2018-04-11T08:22:48.885Z lightningd(123264): lightning_channeld-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Can't send commit: nothing to send
[Wed Apr 11 17:22:49.478 2018] 2018-04-11T08:22:48.885Z lightningd(123264): lightning_channeld-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: peer_out WIRE_SHUTDOWN
[Wed Apr 11 17:22:49.522 2018] 2018-04-11T08:22:48.885Z lightningd(123264): lightning_channeld-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: UPDATE WIRE_CHANNEL_SHUTDOWN_COMPLETE
[Wed Apr 11 17:22:49.522 2018] 2018-04-11T08:22:48.885Z lightningd(123264): lightning_channeld-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: UPDATE WIRE_CHANNEL_SHUTDOWN_COMPLETE
[Wed Apr 11 17:22:49.522 2018] 2018-04-11T08:22:48.886Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: pid 123822, msgfd 17
[Wed Apr 11 17:22:49.522 2018] 2018-04-11T08:22:48.886Z lightningd(123264): 03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: State changed from CHANNELD_SHUTTING_DOWN to CLOSINGD_SIGEXCHANGE
[Wed Apr 11 17:22:49.522 2018] 2018-04-11T08:22:48.906Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: satoshi_out = 0/90000
[Wed Apr 11 17:22:49.522 2018] 2018-04-11T08:22:48.906Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: dustlimit = 546
[Wed Apr 11 17:22:49.533 2018] 2018-04-11T08:22:48.906Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: fee = 183
[Wed Apr 11 17:22:49.533 2018] 2018-04-11T08:22:48.908Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: peer_in WIRE_CLOSING_SIGNED
[Wed Apr 11 17:22:49.548 2018] 2018-04-11T08:22:48.909Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Making close tx at = 0/90000 fee 362
[Wed Apr 11 17:22:49.548 2018] 2018-04-11T08:22:48.909Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Received fee offer 362
[Wed Apr 11 17:22:49.548 2018] 2018-04-11T08:22:48.909Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: ...offer is reasonable
[Wed Apr 11 17:22:49.548 2018] 2018-04-11T08:22:48.909Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: UPDATE WIRE_CLOSING_RECEIVED_SIGNATURE
[Wed Apr 11 17:22:49.548 2018] 2018-04-11T08:22:48.909Z lightningd(123264): 03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Their actual closing tx fee is 362 vs previous 362
[Wed Apr 11 17:22:49.548 2018] 2018-04-11T08:22:48.909Z lightningd(123264): 03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: ... That's same distance to our ideal 121
[Wed Apr 11 17:22:49.554 2018] 2018-04-11T08:22:48.948Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Making close tx at = 0/90000 fee 183
[Wed Apr 11 17:22:49.554 2018] 2018-04-11T08:22:48.948Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: sending fee offer 183
[Wed Apr 11 17:22:49.589 2018] 2018-04-11T08:22:48.948Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: peer_out WIRE_CLOSING_SIGNED
[Wed Apr 11 17:22:49.589 2018] 2018-04-11T08:22:48.948Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Feerange init 0-362, remote higher
[Wed Apr 11 17:22:49.589 2018] 2018-04-11T08:22:48.948Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Feerange local update 183: now 184-362
[Wed Apr 11 17:22:49.589 2018] 2018-04-11T08:22:48.948Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Feerange remote update 362: now 184-361
[Wed Apr 11 17:22:49.589 2018] 2018-04-11T08:22:48.949Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Making close tx at = 0/90000 fee 272
[Wed Apr 11 17:22:49.589 2018] 2018-04-11T08:22:48.949Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: sending fee offer 272
[Wed Apr 11 17:22:49.597 2018] 2018-04-11T08:22:48.949Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: peer_out WIRE_CLOSING_SIGNED
[Wed Apr 11 17:22:49.597 2018] 2018-04-11T08:22:48.949Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Feerange local update 272: now 273-361
[Wed Apr 11 17:22:49.606 2018] 2018-04-11T08:22:48.976Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: peer_in WIRE_CLOSING_SIGNED
[Wed Apr 11 17:22:49.606 2018] 2018-04-11T08:22:48.976Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Making close tx at = 0/90000 fee 183
[Wed Apr 11 17:22:49.606 2018] 2018-04-11T08:22:48.977Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Received fee offer 183
[Wed Apr 11 17:22:49.606 2018] 2018-04-11T08:22:48.977Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: ...offer is reasonable
[Wed Apr 11 17:22:49.606 2018] 2018-04-11T08:22:48.977Z lightningd(123264): lightning_closingd-03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: UPDATE WIRE_CLOSING_RECEIVED_SIGNATURE
[Wed Apr 11 17:22:49.606 2018] 2018-04-11T08:22:48.977Z lightningd(123264): 03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Their actual closing tx fee is 183 vs previous 362
[Wed Apr 11 17:22:49.615 2018] 2018-04-11T08:22:48.977Z lightningd(123264): 03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: ... That's closer to our ideal 121
[Wed Apr 11 17:22:49.615 2018] 2018-04-11T08:22:48.995Z lightningd(123264): 03d09bdb6da742dd1039e1ff40517c86e331011e86fa0dc26bf13e31dc193959cd chan #1: Peer permanent failure in CLOSINGD_SIGEXCHANGE: lightning_closingd: sent ERROR remote offer 183 not between 273 and 361
[Wed Apr 11 17:22:49.630 2018] 2018-04-11T08:22:48.995Z lightningd(123264): 	 (tx e4cef609f9bea5958a387637532c33b10ea7eb126d3a5828042c8101d8e92a60)
[Wed Apr 11 17:22:49.630 2018] 2018-04-11T08:22:48.995Z lightningd(123264): sendrawtransaction: 020000000001010020e0b5ea416d253b242de037e7384f654e2cfd7a41a5b4809a723fcd31c7a20000000000ffffffff01d95e01000000000017a9143601b53ba769b98d43546e5bc8cefc821c2bf364870400483045022100dfa82868cdb5b76513151b13fffa4b43ee65cb69624bd28929685f64599859c8022050e00d33ee221c20311ad6697055d3488baa9d7fe725d4f9adfaecbdd1422a8101483045022100d41e5c85a023b21a6fc04eaff798ae482c019f052836b924019bb0192329322202204da3982768c09746a9f309f232711ef3bfd4a5373a71a05997d649ff4a2d69a401475221020f11eeafb075da68abf82d4c81ae6b2b02a85404165f92d320e09a5671344a9d2103b34e5b61a4302a6ed68775e7f27da1815831a929480419362240e37077bd9fa152ae00000000
```

**Comment by ZmnSCPxj** [comment]
Let me put numbers:

1. ptarmigan: 362
2. clightning: 183
3. clightning: 272 (1 millisecond later!)
4. ptarmigan: 183

It does seem strange that clightning sends two `closing_signed` messages in close succession.

Hmm.  BOLT 2 says:

>Once shutdown is complete and the channel is empty of HTLCs, the final current commitment transactions will have no HTLCs, and closing fee negotiation begins. The funder chooses a fee it thinks is fair, and signs the close transaction with the `scriptpubkey` fields from the `shutdown` messages (along with its chosen fee) and sends the signature; the other node then replies similarly, using a fee it thinks is fair.

Who is the funder of the channel in this case? clightning or ptarmigan?  If clightning is funder then it should have sent the `closing_signed` first, not ptarmigan.

**Comment by ZmnSCPxj** [comment]
> * funder: our node

Ha, sorry it is right there.

Let me see the `closingd` code, looks like a bug on our side.

**Comment by ZmnSCPxj** [comment]
I think I know where the bug is.

The first two feerate exchanges are handled by the `for` loop below: https://github.com/ElementsProject/lightning/blob/957513666c494e35f71cb735eb82faed08f7b594/closingd/closing.c#L498-L499

The side is either `LOCAL` (`0`) or `REMOTE` (`1`).  The `funder` variable is either of those two values.  In the specific case above, `funder = REMOTE` and so the initial `whose_turn` is `REMOTE`.

Inside the `for` loop we check the `whose_turn` variable for whether it is our `LOCAL` turn or their `REMOTE` turn.

Note the `for` loop third clause: `i++, whose_turn = !whose_turn`.  When the loop exits, the counter `i` is 2, and `whose_turn = !!funder` (since we have executed `whose_turn = !whose_turn` twice), meaning `whose_turn` is now `REMOTE` (the original `funder` value).

So: at `i = 0`, `whose_turn = REMOTE`, and we waited for ptarmigan to send its fee.  At `i = 1`, `whose_turn = LOCAL`, and we sent our fee.  Then we did the `for` loop third clause resulting in `i = 2`, `whose_turn = REMOTE`.

Now we continue to the next loop which is the negotiation part: 
https://github.com/ElementsProject/lightning/blob/957513666c494e35f71cb735eb82faed08f7b594/closingd/closing.c#L535-L544

And we see our mistake!https://github.com/ElementsProject/lightning/blob/957513666c494e35f71cb735eb82faed08f7b594/closingd/closing.c#L541-L544

Remember that at this point, `whose_turn` is `REMOTE`, and we just sent our fee.  But now we are flipping `whose_turn` BEFORE we check it, so it flips to `LOCAL` and we send our fee AGAIN!

@rustyrussell to confirm, will make a PR.
